### DISCUSSION TITLE
Making Zeek load scripts by default, outside of explicit `@load`s

### DISCUSSION THREAD

**Discussion by ckreibich** [zeek/zeek#4038] [General]
Zeek currently loads a handful of wired-in files at startup, plus anything provided on the command line (with some additional details, like scripts shipped with plugins etc). I'm not aware of anything that would make it load additional files desired by the user, by default. This would be handy in some settings, for example testing/CI, where you may want to affect Zeek's behavior on a range of tests, or simply as part of your site-local defaults. This therefore also relates to local.zeek, which we point at for local customizations, but which is only loaded because zeekctl tells Zeek to do so.

So I'm wondering how we could do better. A few things come to mind:

- Just add an environment variable that can spell out scripts. Not super elegant, but it would get the job done.
- Expand our concept of the `site/` folder, and start supporting `.d`-style directories (i.e., any Zeek scripts put in that folder get loaded), perhaps `site/autoload.d` or some such.

I'd prefer the latter, since it's more explicit and in line with what many other applications do. But I mainly wanted to make sure I'm not missing prior work or discussion of this. Thoughts welcome...

ps: I once prototyped the directories a bit, but never returned to it: https://github.com/corelight/zeek-globload

**Comment by vpax** [comment]
Well you're likely not going to like this, but that is what `$ZEEK_PREFIXES` was originally introduced to do. Seems that plus a shadow directory tree of (just a few) base scripts in `$ZEEKPATH` might already support this?

**Reply by ckreibich** [reply]
Try me, I have never really understood the prefix mechanism. :-) Can you explain it? I had filed this as an approach to override where specific scripts that would be loaded anyway are loaded from ... is that right? I've never come across actual use of it.

I'm not sure I follow the latter idea — you mean piggy-backing on the scripts that Zeek already auto-loads, by adjusting the location it loads them from?

**Reply by vpax** [reply]
> Try me, I have never really understood the prefix mechanism. :-)

The way it works, IIRC (and I just checked [the docs](https://docs.zeek.org/en/current/script-reference/directives.html#prefixes) to confirm/fix my recollections), if you have a prefix like `test` then for every `foo.zeek` that's loaded, Zeek will (later) also try to load `test.foo.zeek`, with no complaint if it's not found.

This provides a nifty way to add refinement to existing stuff (and come to think of it, at Corelight we should mull over leveraging it more ... assuming it doesn't have bit-rot :P)

**Reply by vpax** [reply]
Oh, and the docs indicate the pathname prefixes get stripped out, so for `policy/foo.zeek` you don't need to provide `policy/test.foo.zeek` somewhere in the path, just `test.foo.zeek`.

**Reply by ckreibich** [reply]
Thanks! Ah right, this works in addition. I don't dislike it for what it provides, but I don't think it addresses the use-case. It does give you a way to tuck on scripts, true. But I'd argue users shouldn't have to think in terms of tweaking scripts already loaded (just thinking of test cases, the most common denominator would likely be `init-bare`, which is a low-level detail of Zeek's bootstrap).

A `.d`-style directory would allow adding additional scripts without needing to know Zeek internals, and do so without administrative coordination, i.e. people wouldn't need to jointly edit a `__load__.zeek` or some such, which aids automation.

> at Corelight we should mull over leveraging it more ...

Hah! :-)

(Btw I have seen this before, in the sense that `-p` was set on Zeek invocations. I have not encountered the mechanism actually used.)

**Reply by timwoj** [reply]
For context, the use-case that drove this entire discussion originally is that there isn't an easy way to say "load this policy file when running all btests", namely the one to remove the `ip_proto` field from conn.log. Either of the two approaches here would accomplish this, and might actually be better than just a list of scripts to automatically load.
